hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellBlocks.java

```diff
@@ -0,0 +1,427 @@
+  @Override 
+  public Entry<Cell, Cell> lowerEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> higherEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> ceilingEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> floorEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> firstEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> lastEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> pollFirstEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> pollLastEntry() { throw new UnsupportedOperationException(); }
+
+
+  // -------------------------------- Updates --------------------------------
+  // All updating methods below are unsupported.
+  // Assuming an array of Cells will be allocated externally,
+  // fill up with Cells and provided in construction time.
+  // Later the structure is immutable.
+  @Override public Cell put(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+  @Override public void clear() { throw new UnsupportedOperationException(); }
+
+  @Override public Cell remove(Object o) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public boolean replace(Cell k, Cell v, Cell v1) { throw new UnsupportedOperationException(); }
+
+  @Override public void putAll(Map<? extends Cell, ? extends Cell> map) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override public Cell putIfAbsent(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public boolean remove(Object o, Object o1) { throw new UnsupportedOperationException(); }
+
+  @Override public Cell replace(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+
+  // -------------------------------- Sub-Sets --------------------------------
+  @Override public NavigableSet<Cell> navigableKeySet() { return new CellBlocksSet(); }
+
+  @Override 
+  public NavigableSet<Cell> descendingKeySet() { throw new UnsupportedOperationException(); }
+
+  @Override public NavigableSet<Cell> keySet() { throw new UnsupportedOperationException(); }
+
+  @Override public Collection<Cell> values() { return new CellBlocksCollection(); }
+
+  @Override public Set<Entry<Cell, Cell>> entrySet() { throw new UnsupportedOperationException(); }
+
+
+  // -------------------------------- Iterator K --------------------------------
+  private final class CellBlocksIterator implements Iterator<Cell> {
+    int index = minCellIdx;
+
+    private CellBlocksIterator(boolean d) {
+//      org.junit.Assert.assertTrue("\nInitializing Cell iterator, descending?:" + d + "\n",false);
+
+      isDescending = d;
```

Comment:
alternative option is to have two classes ascending and descending iterator, but since the code is so small we can leave with the current solution IMO.